### PR TITLE
improve onboarding experiment 1

### PR DESCRIPTION
**Pull request by brindy**
Task/Issue URL: https://app.asana.com/0/392891325557410/1110624620486758
Tech Design URL: n/a
CC:

**Description**:

Shows new onboarding depending on the variant.  See tests for details.

Tests should be performed on a range of devices:

* iPhone SE
* iPhone 6 / 8
* iPhone X
* iPad

Note: ATB tests now pass an environment variable to override the variant for consistency, setting it to `sc`, so that UI tests don't need to be updated for every experiment.

**Steps to test this PR**:

Control group `sc`:
1. Set variant to the control group (or blank)
1. Onboarding should behave as it has done thus far

Summary `mq`:
1. Set the variant to the summary onboarding
1. The onboarding is now a single summary page with a "Starting Browsing" button
1. Tapping the button cross-dissolves to the home screen
1. Rotation not supported for phone devices, but is for iPad devices

Summary + Contextual `mr`:
1. As summary plus...
1. Set the variant to the contextual onboarding
1. The app loads straight in to the home view
1. The first time on the home view a contextual tip showing "search privately" message should show
1. The second time on the home view a contextual tip showing "customize" message should show
1. The first time browsing a none-DDG page the "privacy grade" tip should show 
1. The second time browsing a non-DDG page (could even just be tapping a link on the above page) the "fire button" tip should show
1. Tapping anywhere on the screen should dismiss the tip[1] while simultaneously allowing whatever interaction the user intended to do

[1] A delay of 1 second has been deliberately added to prevent accidental dismissal. 

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: DuckDuckGo/AppDelegate.swift [hunk 1]
```diff
@@ -42,18 +50,28 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         testing = ProcessInfo().arguments.contains("testing")
         if testing {
             window?.rootViewController = UIStoryboard.init(name: "LaunchScreen", bundle: nil).instantiateInitialViewController()
+            return true
```
Comment (brindy) [inline DuckDuckGo/AppDelegate.swift:46]: No need to progress beyond this point for unit tests. This should have been done when this was added.

File: DuckDuckGo/AppDelegate.swift [hunk 2]
```diff
@@ -42,18 +50,28 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
         // assign it here, because "did become active" is already too late and "viewWillAppear"
         // has already been called on the HomeViewController so won't show the home row CTA
         AtbAndVariantCleanup.cleanup()
         DefaultVariantManager().assignVariantIfNeeded()
         HomePageConfiguration.installNewUserFavorites()
-        
-        if !testing {
-            autoClear = AutoClear(worker: mainViewController!)
+
+        if let main = mainViewController {
```
Comment (brindy) [inline DuckDuckGo/AppDelegate.swift:60]: Can't depend on mainViewController being set if app has started in the background.

File: DuckDuckGo/AppDelegate.swift [hunk 3]
```diff
@@ -66,7 +84,6 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         
         Pixel.fire(pixel: .appLaunch)
         StatisticsLoader.shared.load()
-        startOnboardingFlowIfNotSeenBefore()
```
Comment (brindy) [inline DuckDuckGo/AppDelegate.swift:69]: moved to `MainViewController` via an extension in `Onboarding`

File: DuckDuckGo/HomeViewController.swift [hunk 4]
```diff
@@ -107,17 +107,15 @@ class HomeViewController: UIViewController {
         delegate?.showSettings(self)
     }
     
-    override func viewWillAppear(_ animated: Bool) {
-        super.viewWillAppear(animated)
-
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+
```
Comment (brindy) [inline DuckDuckGo/HomeViewController.swift:112]: changed this from `will` to `did` as it makes the animation a bit cleaner and also a better point to hook in the home screen tips

File: DuckDuckGo/Onboarding.swift [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            return
+        } else if variant?.features.contains(.onboardingSummary) ?? false {
+            modalTransitionStyle = .crossDissolve
+            onboardingFlow = isPad ? "OnboardingSummary-iPad" : "OnboardingSummary"
+        } else {
+            modalTransitionStyle = .flipHorizontal
+            onboardingFlow = "Onboarding"
+        }
+        
+        guard let controller = UIStoryboard(name: onboardingFlow, bundle: nil).instantiateInitialViewController() else {
```
Comment (brindy) [inline DuckDuckGo/Onboarding.swift:58]: Do this instead of a segue for finer control.  Same could be achieved from segues, but it means that the `prepareForSegue` code isn't polluted with something potentially temporary.

File: DuckDuckGo/OnboardingSummary-iPad.storyboard [hunk 6]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (brindy) [inline DuckDuckGo/OnboardingSummary-iPad.storyboard:1]: This seemed like an easier way to constrain that view for iPad than trying to do magic with constraints in a single storyboard.

**Comment by brindy** [comment]
(Hopefully) fixed the ATB tests by passing an environment variable to override the variant for consistency, setting it to `sc`, so that UI tests don't need to be updated for every experiment.

File: DuckDuckGo/HomeScreenTips.swift [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    
+}
+
+class HomeScreenTips {
+    
+    enum Tips: Int {
+        case privateSearch
+        case showCustomize
+        
+        static let all = [ privateSearch, showCustomize ]
```
Comment (bwaresiak) [inline DuckDuckGo/HomeScreenTips.swift]: Maybe instead of 'all' make `Tips` a `CaseIterable`?

File: DuckDuckGo/BrowsingTips.swift [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+    private let appUrls = AppUrls()
+    
+    private weak var delegate: BrowsingTipsDelegate?
+    private var storage: ContextualTipsStorage
+    
+    init?(delegate: BrowsingTipsDelegate,
+          storage: ContextualTipsStorage = DefaultContextualTipsStorage(),
+          variantManager: VariantManager = DefaultVariantManager()) {
+        
+        guard (variantManager.currentVariant?.features ?? []).contains(.onboardingContextual) else {
```
Comment (bwaresiak) [inline DuckDuckGo/BrowsingTips.swift]: Maybe we should make some helper method on variant manager (like `isSupported(_ feature: FeatureName)`) to avoid all these optionals and checks?

File: DuckDuckGo/OnboardingSummaryPad.swift [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import UIKit
+
+class OnboardingSummaryPad: UIViewController, Onboarding, OnboardingDelegate {
```
Comment (bwaresiak) [inline DuckDuckGo/OnboardingSummaryPad.swift]: Maybe since we have `OnboardingSummaryViewController` this should be also suffixed with `ViewController`?

File: DuckDuckGo/AppDelegate.swift [hunk 10]
```diff
@@ -42,18 +43,28 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         testing = ProcessInfo().arguments.contains("testing")
         if testing {
             window?.rootViewController = UIStoryboard.init(name: "LaunchScreen", bundle: nil).instantiateInitialViewController()
+            return true
         }
-
+        
+        // only replace an existing window in case starting up for background mode
```
Comment (bwaresiak) [inline DuckDuckGo/AppDelegate.swift]: As we talked about: `if` can most likely be removed.

**Review by bwaresiak — COMMENTED** [review]
Great job @brindy, I really like how easy it is to add new Tips. :)

I've left few comment inline, besides I've found some scenarios that are potential issues:

**Change theme tip**
1. Install App with contextual variant.
2. Open settings and choose light theme. 
3. As a result, tip telling about light theme will be displayed even tho light theme is already set.

**Tips hidden by other views**
One:
1. Install App with contextual variant.
2. Open webpage.
3. Open tabs list, then remove current tab.
4. Settings tip is presented underneath tabs screen and is not accessible. 

Two (more likely to happen in case user has slow connection):
1. Install App with contextual variant.
2. Open webpage.
3. Before page is loaded open tabs list.
4. Grade icon tip is presented underneath tabs screen and is not accessible. 

**Rotation issues on iPad.**
There are issues around rotation on iPad, especially visible around fire button and settings tips.

**Comment by brindy** [comment]
@bwaresiak that's ready for re-review when you get chance, please.

We're going to ignore the edge-case behaviours you found for this experiment. Good spots though!

I can't work out what's wrong with BitRise - it'll build develop, but not this branch. Please run the unit and atb tests locally to be sure.

Thanks!

**Review by bwaresiak — APPROVED** [review]
Good job, works really well!
